---
title: "Further examples"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Further examples}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

<style>
h2,h3,h4 { margin-top: 200px; }
</style>
<script>

function bindButton(widget, button, state) {
    document.getElementById(button).onclick = function() {
        document.getElementById(widget).langevitour.setState(state);
    }
}

</script>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(langevitour)
```

## Mouse brain cells

Single-cell RNA-Seq gene expression of 2,816 mouse brain cells ([Zeisel, 2015](https://pubmed.ncbi.nlm.nih.gov/25700174/)). The top 10 principal components were produced using the steps in the [Bioconductor OSCA workflow](http://bioconductor.org/books/3.13/OSCA.workflows/zeisel-mouse-brain-strt-seq.html). This is unscaled PCA of log$_2$(normalized_count+1) expression levels, mean UMI count 15,550 per cell. So $x$ units of distance can be explained for example by $n$ genes with $x/\sqrt{n}$ log$_2$ fold change each.

```{r}
data(zeiselPC)

palette <- rainbow(7, v=0.75)

langevitour(
    zeiselPC[,-1], zeiselPC$type, 
    levelColors=palette, elementId="brain1")
```

<br>

* Try setting the guide to <button id=brain1Local>local</button>.

* Examine particular cell types in detail by <button id=brain1Examine>turning off</button> the other cell types.

* Get a 3D effect by <button id=brain1ThreeAxes>showing only three axes</button>.

<button id=brain1Reset>Reset</button>

<script>
bindButton('brain1','brain1Local', {"guideType": "local"});

bindButton('brain1','brain1Examine', {
    "guideType": "local",
    "labelInactive": [
        "astrocytes_ependymal",
        "endothelial-mural",
        "microglia",
        "oligodendrocytes"
    ],
});

bindButton('brain1','brain1ThreeAxes', {
    "guideType": "none",
    "labelInactive": ["PC4","PC5","PC6","PC7","PC8","PC9","PC10"],
});

bindButton('brain1','brain1Reset', {
    "guideType": "none",
    "labelInactive": [ ],
});
</script>



## Mouse brain cells denoised

The single cell data is noisy. Typically this sort of data would be examined using a UMAP or t-SNE layout. The UMAP layout below looks very clean. Can we achieve something similar?

```{r echo=FALSE,message=FALSE}
library(uwot)
library(ggplot2)

layout <- umap(zeiselPC[,-1])
ggplot() + 
    aes(layout[,1],layout[,2], color=zeiselPC$type) + 
    geom_point(size=0.5) + 
    labs(x="",y="",color="",title="UMAP layout") + 
    coord_fixed() + 
    scale_x_continuous(breaks=c()) + 
    scale_y_continuous(breaks=c()) +
    scale_color_discrete(type=palette) +
    guides(colour = guide_legend(override.aes = list(size=5))) +
    theme_bw()
```

My current suggestion to reduce noise is, for each point, to take the average of cells reachable in a certain number of steps along the directed k-nearest neighbor graph. See the function `knnDenoise()`.

```{r}
zeiselDenoised <- knnDenoise(zeiselPC[,-1], k=30, steps=2)

langevitour(
    zeiselDenoised, zeiselPC$type, 
    levelColors=palette, elementId="brain2")
```

<br>

* Try setting the guide to <button id=brain2UltraLocal>ultralocal</button>. Ultralocal mode may need some fine-tuning of settings to work nicely, such as turning down the heat.

* Notice that interneurons and the pyramidal CA1 and SS neurons parallel each other. This was not clear in the UMAP layout above.

* Construct a <button id=brain2Manual>manual layout</button> separating all the clusters by dragging PC labels onto the plot. (<button id=brain2Reset>reset</button>)

<script>
bindButton('brain2','brain2UltraLocal', {
    "guideType": "ultralocal", 
    "heat": -1,
    "labelPos": [ ],
});

bindButton('brain2', 'brain2Manual', {
    "guideType": "none",
    "labelAttractionOn": true,
    "labelAttraction": 0,
    "labelInactive": [],
    "labelPos": {
        "PC1": [
            0.4171717171717171,
            0.8959595959595958
        ],
        "PC2": [
            -0.8313131313131312,
            0.44949494949494945
        ],
        "PC3": [
            -0.36868686868686873,
            -0.7343434343434343
        ],
        "PC4": [
            0.6151515151515152,
            -0.25151515151515147
        ],
        "PC5": [
            0.603030303030303,
            -0.03737373737373728
        ],
        "PC6": [
            -0.8090909090909091,
            0.5565656565656565
        ],
        "PC7": [
            -0.6313131313131313,
            -0.009090909090909038
        ],
        "PC8": [
            -0.24901960784313726,
            -0.19078428979013484
        ],
        "PC9": [
            0.4575757575757575,
            -0.31414141414141405
        ]
    },
});

bindButton('brain2', 'brain2Reset', {
    "labelPos": { },
});
</script>


## Mouse brain cells denoised vs original

The denoised version looks much better, but is it a faithful representation of the data? 

UMAP and t-SNE can produce arbitrary distortions and ripping apart of nearby points. Langevitour will only ever present linear projections of data, avoiding this problem. We should have a faithful view of the data so long as our denoising step is also relatively faithful to the original.

Let's compare the original and denoised versions. You can judge for yourself if the original geometry is faithfully retained. 

```{r}
n <- nrow(zeiselPC)
lineFrom <- seq_len(n)
lineTo <- lineFrom + n

langevitour(
    rbind(zeiselPC[,-1], zeiselDenoised), 
    rep(c("Original","Denoised"), each=n), 
    lineFrom=lineFrom, lineTo=lineTo)
```


## Olive oils

A dataset from the [tourr](http://ggobi.github.io/tourr/) package.

By default, langevitour does not scale variables individually. This dataset rather requires scaling, so we supply our desired scaling.

I also demonstrate adding extra axes representing principal components.

```{r}
olives <- tourr::olive[,3:10]
olivesScale <- apply(olives, 2, sd)
olivesGroup <- paste(tourr::olive$region, tourr::olive$area)

# Find principal components
pca <- prcomp(sweep(olives, 2, olivesScale, "/"), rank=2)
olivesPC <- sweep(pca$rotation, 1, olivesScale, "/")

langevitour(
    olives, 
    group=olivesGroup, 
    scale=olivesScale*6, 
    extraAxes=olivesPC,
    pointSize=1.5)
```


## Hyper-cube

Some geometry from the [Geo Zoo](https://github.com/schloerke/geozoo).

```{r}
cube <- geozoo::cube.solid.grid(5, 1)
langevitour(
    cube$points, 
    lineFrom=cube$edges[,"from"], 
    lineTo=cube$edges[,"to"], 
    lineColors=rainbow(nrow(cube$edges)),
    pointSize=sqrt(seq_len(nrow(cube$points))))
```

## Torus

```{r}
langevitour(geozoo::torus.flat(p=4)$points)
```

## pdfsense

A dataset in the [liminal](https://sa-lee.github.io/liminal/) package. Each row has been scaled to a unit vector.

```{r}
X <- as.matrix(liminal::pdfsense[,-(1:6)])
X <- X / sqrt(rowSums(X*X))
langevitour(X, liminal::pdfsense$Type)
```

<br>

* Try setting the guide to "outliers".


## Stiefel manifold $V_a(\mathbb{R}^b)$

The orthonormal projections of b dimensions onto a.

Stiefel manifolds are what langevitour tours!

```{r}
stiefel <- function(a,b) {
    k <- 5000
    
    X <- matrix(nrow=k, ncol=a*b)
    colnames(X) <- paste0("X[",rep(seq_len(a),b),",",rep(seq_len(b), each=a),"]")
    for(i in seq_len(k)) {
        mat <- matrix(rnorm(a*b), nrow=a)
        decomp <- svd(mat)
        mat <- decomp$u %*% t(decomp$v)
        X[i,] <- as.vector(mat)
    }
    
    langevitour(X)
}

stiefel(2,3)
```

```{r}
stiefel(2,2)
```